### PR TITLE
media rating sync

### LINKED ISSUE #1678: ID3 rating tag sync
It would be great to have functionality to read and write to the id3 tag with metadata from owntone. I'm currently using PlexAmp to sync music to my phone and listen when I'm not at home, and I have that set up to sync the rating with the file's id3 tag - I believe it's using the python mutagen library.

It would be amazing to have the ability to read and write the ratings to the files' id3 tags. I've lost all my metadata from my itunes days, and my playlists based on ratings, dates added, played, skipped, and play/skip count are all now starting over from scratch. I would love the ability to write this data directly to the file metadata, so that I can keep all that in sync across my devices and platforms.

Alternatively, a sync solution for android would solve all my problems! :)

### PR DESCRIPTION
**Pull request by whatdoineed2do**
Closes #1678

* read `rating` from local media on scan
* new library api to automatically update local files when rating changes

Test:
```
# create dummy test files
cd /svr/Music
ffmpeg -f lavfi -i "sine=frequency=500:duration=5" -af "afade=t=out:st=3:d=2" -ac 2 -ar 48000 foo.mp3
ffmpeg -f lavfi -i "sine=frequency=500:duration=5" -af "afade=t=out:st=3:d=2" -ac 2 -ar 48000 foo.flac
ffmpeg -f lavfi -i "sine=frequency=500:duration=5" -af "afade=t=out:st=3:d=2" -ac 2 -ar 48000 bar.mp3
ffmpeg -f lavfi -i "sine=frequency=500:duration=5" -af "afade=t=out:st=3:d=2" -ac 2 -ar 48000 bar.m4a

# wait for server to find files

# edge case test
chmod 000 foo.mp3

# set rating, m4a will not  work
# will see in logs:
#     scan: Updating rating to 60 on '/.../bar.mp3'
#     scan: No permissions to update metadata on '/.../foo.mp3' - skipping
#     scan: Unsupported metadata (aac) update for 'rating' on '/.../bar.m4a' (aac) - skipping
#     scan: Unsupported media (url) to update metadata on 'https://....' (id=1251)
#     scan: No known DB entry for media, '99999' to update metadata
IDS=$(sqlite3 /.../songs.db  'select id from files where path like "%bar.%" or path like "%foo.%"')

for ID in ${IDS}; do 
  curl -XPUT http://localhost:3689/api/library/tracks/${ID}?rating=60
done

# verify rating value
for i in foo.mp3 foo.flac bar.mp3 bar.m4a; do ffmpeg -hide_banner -i $i; done

# verify server inotifications still work, will see usual "cache: DAAP cache updated"
touch bar.mp3

# verify handling of non local streams - will see in log:  scan: Ignoring non local media (...) requested for rating write
ID=$(sqlite3 /.../songs.db  'select id from files where path like "http%" limit 1')

curl -XPUT http://localhost:3689/api/library/tracks/${ID}?rating=60
```

### CODE REVIEW

**Comment by whatdoineed2do** [comment]
@ejurgensen - can you provide feedback on the inotify handling part?  This seems to do what we want with specific inotify removal per file being updated.


I did find an (odd?) situation with that my testing db `inotify` tbl has less lines than my `files` tbl and i dont know what causes this or if this is normal.

**Comment by ejurgensen** [comment]
I have received a row of notifications for amendments to this PR, please don't create PRs for work in progress. It's of course fine to ask for feedback on a draft, but then mark it as a draft and continue working on another branch.

**Comment by whatdoineed2do** [comment]
i'm done with changes

**Comment by whatdoineed2do** [comment]
One more fix for inotify removal but this should be ready for review

**Comment by ejurgensen** [comment]
I still don't understand the purpose of the extra endpoint? Is it because you don't want to write the metadata automatically?

**Comment by whatdoineed2do** [comment]
Yes, didnt want auto update.  Auto updating at the db update seems too low level to also manipulate the media file esp if there are errors to  report.

Also it may not be something a user wants or enables if media is on a read only filesystem (NFS mount like one of my RPis) so I choose the option that required a user request for this to happen.

We _could_ use the existing  endpoint which accepts rating updates in `jsonapi_reply_library_tracks_put_byid()` when a user can force an rating update but this makes it asymetrical for aglorithmic rating updates - ie the media file still becomes out of snyc with what the server/db maintains.

**Comment by ejurgensen** [comment]
I think the sync should be automatic, but with a setting/config option that makes it disabled by default. The implementation that does the update should be in the library layer (either an extension of filescanner_ffmpeg or separate module), so not in the API.

**Comment by whatdoineed2do** [comment]
Let me know if this is more what you're after.

The rating update to the media file happens after the db rating is applied and iterates for all the library sources asking them to sync metadata, where (atm) only file scanner implements this.  All ffmpeg related items are in filescanner_ffmpeg

I'll cleanup the commits if you're ok with this approach.

To test, find an id of the local supported media file (mp3 or flac):
```
curl -XPUT http://localhost:3689/api/library/tracks/1239?rating=60
```

I haven't quited tested the mpd auto-rating update but its through the same path using the virtual_path supplied

**Comment by ejurgensen** [comment]
Yes, this structure looks good. Have you looked into if there are any interfaces that update the rating via db_file_update()? I guess that would be challenging to catch and sync.

**Comment by whatdoineed2do** [comment]
`db_file_update()` used in 
* spotify
* rss scanner
* itunes
* REST endpoint

Looking at these it doesnt makes sense for the remote items (top 3 in list) to try to update remote media.

The REST endpoint may be the only one of interest (`jsonapi_reply_library_tracks_put()` `library_media_save()` -> `db_file_update()`) so i've kept it out of the db layer and left that change to the library layer with an additional commit - looking at the web ui, i dont see rating udpates being sent this way (at the moment)

**Comment by whatdoineed2do** [comment]
@ejurgensen - any more thoughts on this PR? Otherwise I'll squash the commits for the merge

**Comment by ejurgensen** [comment]
Overall it looks fine, but I need to find time for a more detailed review. You're welcome to squash.

File: src/library/filescanner.c [hunk 1]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+  dirname(inotify_path);
+
+  if (access(path, W_OK) < 0 || access(inotify_path, W_OK) < 0)
+    {
+      DPRINTF(E_INFO, L_SCAN, "No permissions to update metadata, skipping %s\n", path);
+      return 0;
+    }
+
+  // Temporarily disable inotify
+  ret = db_watch_get_bypath(&wi, inotify_path);
```
Comment (ejurgensen) [inline src/library/filescanner.c:1795]: Looks like watches_clear() also clears all subdirs? Might be better to just make a watch_rm and watch_add

File: src/library/filescanner.c [hunk 2]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+      free_wi(&wi, 1);
+    }
+
+
+  filescanner_ffmpeg_sync_metadata(path, rating);
+
+
+  // and re-enable
+#ifdef __linux__
+  flags = IN_ATTRIB | IN_CREATE | IN_DELETE | IN_CLOSE_WRITE | IN_MOVE | IN_DELETE | IN_MOVE_SELF;
```
Comment (ejurgensen) [inline src/library/filescanner.c:1808]: Code duplication. Could be avioded with a conditional #define in the top.

File: src/library/filescanner_ffmpeg.c [hunk 3]
```diff
@@ -173,6 +174,28 @@ parse_albumid(struct media_file_info *mfi, const char *id_string)
...
 
+static int
+parse_rating(struct media_file_info *mfi, const char *rating_string)
+{
+  uint32_t *rating = (uint32_t *) ((char *) mfi + mfi_offsetof(rating));
+
+  if (safe_atou32(rating_string, rating) != 0)
+    return 0;
+
+  // adjust for internal rating 0-100 range redistributing heurisstically
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:185]: I don't understand this

File: src/library/filescanner_ffmpeg.c [hunk 4]
```diff
@@ -778,3 +802,230 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
 
   return 0;
 }
+
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:805]: Checkpoint - haven't reviewed the below

File: src/library/filescanner.c [hunk 5]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+      DPRINTF(E_LOG, L_SCAN, "Bug! Invalid params %s:%d\n", __FILE__, __LINE__);
+      return -1;
+    }
+
+  if (virtual_path)
+    {
+      path = (char*)virtual_path_to_path(virtual_path);
+      if (!path)
+	{
+	  DPRINTF(E_INFO, L_SCAN, "Ignoring non local media %s requested for metadata sync\n", virtual_path);
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1758]: The update is automatic in 2 potential places - the auto-rating calculation BUT also in the potential web ui request (`http://localhost:3689/api/library/tracks/1239?rating=40`) REST endpoint;

File: src/library/filescanner.c [hunk 6]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
   return 0;
 }
 
+static int
+filescanner_sync_metadata(const char *virtual_path, const uint32_t *id, uint32_t rating)
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1737]: no, this is called in 2 places.  the REST endpoint `jsonapi_reply_library_tracks_put_byid()` which does not have the `mfi` at that point.

File: src/library/filescanner.c [hunk 7]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+	  DPRINTF(E_LOG, L_SCAN, "No known path for local media id=%d\n", *id);
+	  return -1;
+	}
+      strcpy(id_path, path);
+      free(path);
+
+      path = id_path;
+    }
+
+  // Inotify watches dir paths
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1784]: in my `inotify` table, all paths are directories - therefore i need to get the dir path, since the inotify backend "watches" directory paths only.  does that clear it up?

File: src/library/filescanner.c [hunk 8]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+  strcpy(inotify_path, path);
+  dirname(inotify_path);
+
+  if (access(path, W_OK) < 0 || access(inotify_path, W_OK) < 0)
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1788]: This is quite expensive to not have this check so i'd urge to  keep it as it allows generating the new file with the updated data ONLY to know already that we can't actually make the update.

File: src/library/filescanner_ffmpeg.c (hunk 3 above)
Comment (whatdoineed2do) [reply src/library/filescanner_ffmpeg.c:185]: Some taggers set rating in rnage 0..255 - this code adjusts the file rating to 0..100 that the server uses

File: src/library.c [hunk 9]
```diff
@@ -134,9 +136,14 @@ library_media_save(struct media_file_info *mfi)
     }
 
   if (mfi->id == 0)
-    return db_file_add(mfi);
+    ret = db_file_add(mfi);
   else
-    return db_file_update(mfi);
+    {
+      ret = db_file_update(mfi);
+      if (mfi->data_kind == DATA_KIND_FILE)
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library.c]: uses the library sources array to let the enabled/implemented callbacks to handle

File: src/library/filescanner_ffmpeg.c [hunk 10]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+  }
+
+  if (!supported)
+    goto end;
+
+  safe_snprintf_cat(rating, 4, "%d", mfi->rating);
+
+  // Save a potential write if metadata on the underlying file matches requested rating
+  entry = av_dict_get(ctx->metadata, "rating", NULL, 0);
+  if (entry == NULL || (entry && entry->value == NULL) || (entry && strcmp(entry->value, rating) != 0) )
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:1007]: Invert this, i.e. return if rating is equal. Then less nesting is required.

File: src/library/filescanner_ffmpeg.c [hunk 11]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+	  if (ret < 0)
+	    {
+	      DPRINTF(E_LOG, L_SCAN, "failed to replace library rating file '%s' with temp '%s' - %s\n", ctx->url, dest, strerror(errno));
+	      unlink(dest);
+	      ret = EACCES;
+	    }
+	}
+      else
+	{
+	  unlink(dest);
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:1027]: But dest may never have been created? Let file_clone_metadata() clean up itself if it fails.

File: src/library/filescanner_ffmpeg.c [hunk 12]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+  char dest[PATH_MAX];
+  int i;
+  bool supported = false;
+  AVDictionaryEntry  *entry;
+
+  AVFormatContext *ctx = NULL;
+  if ( (ret = avformat_open_input(&ctx, mfi->path, NULL, NULL)) != 0)
+    {
+      DPRINTF(E_LOG, L_SCAN, "failed to open library file for rating metadata update '%s' - %s\n", mfi->path, av_err2str(ret));
+      return ENOENT;
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:977]: Positive return values on error?

File: src/library/filescanner_ffmpeg.c [hunk 13]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
 
   return 0;
 }
+
+static int
+file_clone_metadata(AVFormatContext* in_fmt_ctx, const char* dest)
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:807]: Doesn't this function actually clone the entire file, not just the metadata?

File: src/library/filescanner_ffmpeg.c [hunk 14]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+  // Save a potential write if metadata on the underlying file matches requested rating
+  entry = av_dict_get(ctx->metadata, "rating", NULL, 0);
+  if (entry == NULL || (entry && entry->value == NULL) || (entry && strcmp(entry->value, rating) != 0) )
+    {
+      av_dict_set(&ctx->metadata, "rating", rating, 0);
+      DPRINTF(E_LOG, L_SCAN, "updating rating to %s on '%s'\n", rating, mfi->path);
+
+      sprintf(dest, "%s-%d.metadata", ctx->url, getpid());
+
+      if ( (ret = file_clone_metadata(ctx, dest)) == 0)
```
Comment (ejurgensen) [inline src/library/filescanner_ffmpeg.c:1014]: ret is never used. And again, inverting would avoid nesting.

**Review by ejurgensen — COMMENTED** [review]
Here's some more comments. If I understand the implementation correctly, it involves rewriting the entire file, which would seem a bit risky. Is that the only way to save metadata with ffmpeg?

**Comment by whatdoineed2do** [comment]
> Here's some more comments. If I understand the implementation correctly, it involves rewriting the entire file, which would seem a bit risky. Is that the only way to save metadata with ffmpeg?

Yes, this appears to be correct.  not seen any code to suggest otherwise and its how ffmpeg command line does it too (separate input and output files)

File: src/library/filescanner_ffmpeg.c [hunk 15]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
 
   return 0;
 }
+
+static int
+file_clone_metadata(AVFormatContext* in_fmt_ctx, const char* dest)
+{
+  int  ret = 0;
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner_ffmpeg.c]: init to sensible values at point of declaration are not useless.  we have to do them later to the same values regardless but as you wish.  Assembly from gcc suggests init and later assign generate the same instructions anyway

codeql reminds us that the `end` label clean up items and relies on init to NULL etc.  otherwise we'll have to add multiple labels and gotos which seems worse than initialise at point of declaration so i've left a commit with those advised by codql

File: src/library/filescanner_ffmpeg.c [hunk 16]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+      /* copy packet */
+      pkt.pts = av_rescale_q_rnd (pkt.pts, in_stream->time_base, out_stream->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+      pkt.dts = av_rescale_q_rnd (pkt.dts, in_stream->time_base, out_stream->time_base, AV_ROUND_NEAR_INF | AV_ROUND_PASS_MINMAX);
+      pkt.duration = av_rescale_q (pkt.duration, in_stream->time_base, out_stream->time_base);
+      pkt.pos = -1;
+
+      ret = av_interleaved_write_frame (out_fmt_ctx, &pkt);
+      if (ret < 0)
+	{
+	  DPRINTF(E_LOG, L_SCAN, "Error muxing pkt for rating '%s' - %s\n", in_fmt_ctx->url, av_err2str(ret));
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner_ffmpeg.c:941]: yes in failure case of write_frame.  `unref` added before test on `ret`

File: src/library/filescanner_ffmpeg.c [hunk 17]
```diff
@@ -778,3 +802,234 @@ scan_metadata_ffmpeg(struct media_file_info *mfi, const char *file)
...
+  safe_snprintf_cat(rating, 4, "%d", mfi->rating);
+
+  // Save a potential write if metadata on the underlying file matches requested rating
+  entry = av_dict_get(ctx->metadata, "rating", NULL, 0);
+  if (entry == NULL || (entry && entry->value == NULL) || (entry && strcmp(entry->value, rating) != 0) )
+    {
+      av_dict_set(&ctx->metadata, "rating", rating, 0);
+      DPRINTF(E_LOG, L_SCAN, "updating rating to %s on '%s'\n", rating, mfi->path);
+
+      sprintf(dest, "%s-%d.metadata", ctx->url, getpid());
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner_ffmpeg.c]: we do a `rename` later to replace the original file, which means it needs to be on the same device.  the `mkstemps()` template creates a path in the same place as the original file to avoid this limitation from `rename()`

File: src/library/filescanner_ffmpeg.c (hunk 13 above)
Comment (whatdoineed2do) [reply src/library/filescanner_ffmpeg.c]: yes full clone.  renamed

File: src/library/filescanner.c [hunk 18]
```diff
@@ -1719,6 +1733,99 @@ filescanner_fullrescan()
...
+  char *path;
+  int id_kind = 0;
+
+  if (virtual_path == NULL && id == NULL)
+    {
+      DPRINTF(E_LOG, L_SCAN, "Bug! Invalid params %s:%d\n", __FILE__, __LINE__);
+      return -1;
+    }
+
+  if (virtual_path)
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1735]: this gets called by: `mpd.c` and `httpd_{dacp,jsonapi.c` where the mfi is not available - getting it seems a heavy operation - as the client only has the `item id` or `virtual path`

if you want i can change `db_file_rating_update()` to get the `mfi` before handing it to `write_metadata()` but it seems expensive when the db update on the rating is alrady done

[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/library/filescanner.c:1735]: i think i;ve addressed the other items in the review andjust this one.  if i've missed anything let me know

**Comment by whatdoineed2do** [comment]
@ejurgensen  - let me know if you have any additional comments on the review fixes, otherwise i'll squash the commits in prep for merge

**Comment by ejurgensen** [comment]
You're welcome to do that. I will go through it again and just make final edits myself before merging.

**Comment by ejurgensen** [comment]
I started going through this, but there are a few things in filescanner_ffmpeg_write_rating() I couldn't understand:

1. There is check that the format supports metadata updates, but the actual implementation doesn't check format (container), it checks codecs. That seems strange. What's the reason for that? Did you take this check from somewhere?
2. A few lines after, safe_snprintf_cat() is used, but not for concatenation, since rating is just an empty string at that point. Why isn't snprintf used?
3. Why isn't the temp file being stored in /tmp?

You don't need to make any fixes concerning these, I'm fixing it up anyway. Just need to understand the above.

**Comment by whatdoineed2do** [comment]
>     1. There is check that the format supports metadata updates, but the actual implementation doesn't check format (container), it checks codecs. That seems strange. What's the reason for that? Did you take this check from somewhere?
> 

 do it based on the audio codec  but this is prob suboptimal to check against codec vs container - however neither `AVFormatContext` and `AVInputFormat` give a what appears to be clean what to determine container format other than C-strings unless I've missed something on the libavformat api that could do this cleaner.

>     2. A few lines after, safe_snprintf_cat() is used, but not for concatenation, since rating is just an empty string at that point. Why isn't snprintf used?

yes, this looks like an oversight when `sprintf` will do

> 
>     3. Why isn't the temp file being stored in /tmp?

I am using ffmpeg to write the file and then using `rename(2)` to _replace_ the original file - the problem with `rename()` is that it expects the file move to be on the same device/filesystem, and `/tmp` and the library is not necessarily the case - creating file `/tmp` leads to `EXDEV` which would need me to manually byte copy a file to the target location: example below:

```scan: Failed to replace library rating file '/export/public/Music/test/bar.mp3' with temp '/tmp/PVGFYw.metadata' - Invalid cross-device link```

**Comment by ejurgensen** [comment]
Thanks for the info. Locally, I’ve made the adjustments that I want to include before merging (plus fixed the conflicts), but how/when to trigger the write to the files is giving me headaches. I’m not happy with the PR's hooks in db_file_rating_update() and library_media_save(). In additon to being somewhat ugly, I think the hook in db_file_rating_update means that the write will be carried out in the httpd and mpd threads? If so, that’s no good.

A better solution should also work when the rating updates happen via db_file_inc_playcount etc, so that there is no chance that the ratings in the db and in the file diverge.

However, all the alternatives that I can think of also have their problems, so for now I have to think some more about it.

**Comment by ejurgensen** [comment]
I've pushed my changes to the PR, so you can check them. I couldn't find a way to always keep db ratings and file ratings in sync. Maybe it's all the same, because the automatic ratings can potentially adjust ratings of all files in a playlist, which is probably a bit too much to write.

I noticed that updated files don't have the same owner and permissions as the original, and I don't think that's ok. Can you try to find a solution for that? I also noticed that my updated file was smaller than the original, which concerns me a bit. Is something being discarded? I couldn't see what that might be, though.

**Comment by whatdoineed2do** [comment]
> I noticed that updated files don't have the same owner and permissions as the original, and I don't think that's ok. Can you try to find a solution for that?

Let me take a look over the next few days.  Unless we're running as `root` this might not be fixable (any one can add files to the library) - there was a check on the ownership of the file at the start of the rating update method so the server `euid` doesn't have write permissions, it bails out but yes file mask needs to check.

> I also noticed that my updated file was smaller than the original, which concerns me a bit. Is something being discarded? I couldn't see what that might be, though.

I seem to remember fixing a bug in `TagLib` a few yrs ago around metadata, depending on what wrote the original, coulld have metadata overallocated (a hole of additional updates instead need to rewrite the entire file)  - perhaps this is what is being clawed back by ffmpeg?

The most important test for me was validating that the audio stream hash is identical between the original and the new file;  this was done via: `ffmpeg -hide_banner -i foo.mp3 -c:a copy -bsf:a null -f hash -`.  Using `exiftool -v3` will also display the file offsets of various blocks that might be able to show anything odd.

**Comment by whatdoineed2do** [comment]
rebased with master and updated for ownership/file mask although this is done as a _best efforts_ ...

To ensure the ownership of the replacement file is correct, the server needs to be running as `root` but the install/setup seem to default to `owntone` (as per config) and the server drops privs to the config value as it starts even if started as `root`.  Regardless, i've tested this with the `general.uid = root` and works as expected

**Comment by ejurgensen** [comment]
Looks a bit too hacky to me, and like you say, it won't work in some cases. There must be a cleaner way of doing this.

**Comment by whatdoineed2do** [comment]
The alternative to `chmod` is to set `umask ` to match st_mode ahead of `mkstemp` but the ownership item will still be a problem (can anyone other than root change ownership anyway?)

I went with the commit approach to keep the matching logic together.

**Comment by ejurgensen** [comment]
I suggest you look into how a normal tagging app like kid3 does this. Or I guess any app that edits a file. I can't imagine that they mess around with chown/chmod.

Btw, start by just sharing possible ideas. Don't push anything to the branch until we have agreed.

**Comment by whatdoineed2do** [comment]
No unfortunately, we have two options already outlined since there are no other unixy provisions for what you're requiring:

* file mask - set `umask(2)` ahead of the file creation or `chmod(2)`
* file ownership - process must have effective uid as `root` and then we must `chown(2)` the resulting file - this is the ONLY way to change ownership of file by the process; there is no system call available that lets us create a file 'as another user'.
 
To verify this, you can see what mirroring utils (rsync/tar/install as root) do: `mkdir /tmp/dir ; touch /tmp/foo.txt ; strace install -owner nobody /tmp/foo.txt /tmp/dir` - the final operations are a `lchown()` followed by a `chmod()`

I would take this approach and match most utils in unix: _if you copy a file under unix, the resulting file will belong to you_ - even as superuser (excluding _mirroring_ `rsync`, `tar`, `install`..) the duplication makes no provisions to reflect ownership of original (or even file mask).  We should certainly ensure the filemask (on fedora, root user creates files as 0600) which means we replace the last commit with
```
stat(original_file, &ts);
orig_mask = umask(st.st_mode & 0777);
mkstemp(...);
umask(orig_mask);
```
and not do anything with ownership unless we go with more or less what is already in the last commit.

**Comment by whatdoineed2do** [comment]
actually, `fchmod()` is a little cleaner than `umask()` that gives potential race condition:
```
  fd = mkstemp(new_rating_file);
  if (fd < 0)
    {
      DPRINTF(E_WARN, L_SCAN, "Failed to create temp rating file '%s': %s\n", new_rating_file, strerror(errno));
      ret = -1;
      goto end;
    }

  // best efforts: use existing mode, otherwise something sensible and if that
  // fails, it'll be the process's _current_ umask
  fchmod(fd, stat(in_fmt_ctx->url, &st) == 0 ? st.st_mode & 0777 : 0644);

  ret = file_write_rating(in_fmt_ctx, new_rating_file);
```
ownership change still problem for outlined previous

**Comment by whatdoineed2do** [comment]
@ejurgensen - any conclusive thoughts on this based on my comments above:

The additional commit would be as follows
```C
@@ -20,6 +20,7 @@
 # include <config.h>
 #endif
 
+#include <sys/stat.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -956,6 +957,7 @@ write_metadata_ffmpeg(struct media_file_info *mfi)
   int max_rating;
   int file_rating;
   int fd = -1;
+  struct stat st;
   int ret;
 
   // ffmpeg's metadata update are limited - some formats do not support rating
@@ -1008,6 +1010,14 @@ write_metadata_ffmpeg(struct media_file_info *mfi)
       goto end;
     }
 
+  ret = stat(in_fmt_ctx->url, &st);
+  if (ret == 0)
+    {
+      // best efforts...
+      fchmod(fd, st.st_mode & 0777);
+      fchown(fd, st.st_uid, st.st_gid);
+    }
+
   ret = file_write_rating(in_fmt_ctx, new_rating_file);
   if (ret < 0)
     {
```

**Comment by ejurgensen** [comment]
I'll have a look at it

**Comment by ejurgensen** [comment]
I think the way to go about this is to just write directly to the media file. So abandon writing to a copy and so also abandon renaming/replacing/chmod-ing. It will require an extra file copy, because the source file first has to be copied to tmp file to be read from, but I think it will be more clean. I will add a commit to the PR with this.

**Comment by ejurgensen** [comment]
I've made an update to the PR with the different method of updating the file. This method should also mean that the file creation time remains correct, which was also an issue with the other method.

The disabling of inotify might also not be necessary any more.

I won't merge right now, I will be after the next release, so in a few weeks.

**Comment by whatdoineed2do** [comment]
i've added one more commit to this - the `sendfile()` copies bytes but as you observed previosuly the resulting new file may be _smaller_ size.  the `ftruncate()` ensures theres not junk at the in place overwritten file

**Comment by ejurgensen** [comment]
It's not really required with the changed method, since fast_copy() isn't overwriting anything, it's just making a copy of the source. Then ffmpeg does the overwriting from that copy. However, maybe I will change it so that it's ffmpeg first, and then have fast_copy() overwrite to the file, and then ftruncate() will indeed be needed.

**Comment by ejurgensen** [comment]
Squashed and merged now, thanks for the contribution. As you might notice, I removed the disabling of inotify, since with the current solution it isn't really needed. Of course, the file will now be rescanned when the rating is written, but there won't be a loop.
